### ISSUE TITLE
Optim does not handle cloned weights

### ISSUE THREAD

**Issue by douwekiela** [torch/optim#73]
While playing with the example for `MarginRankingCriterion` at https://github.com/torch/nn/blob/master/doc/criterion.md#nn.MarginRankingCriterion, I noticed that optim does not seem to be able to handle cloned weights and biases, due a size mismatch of the flattened parameters and flattened gradient params. Here's a simple example:

```
require 'nn'
require 'optim'

p1_mlp = nn.Linear(5, 2)
p2_mlp = p1_mlp:clone('weight', 'bias')

prl = nn.ParallelTable()
prl:add(p1_mlp)
prl:add(p2_mlp)

mlp1 = nn.Sequential()
mlp1:add(prl)
mlp1:add(nn.DotProduct())

mlp2 = mlp1:clone('weight', 'bias')

mlpa = nn.Sequential()
prla = nn.ParallelTable()
prla:add(mlp1)
prla:add(mlp2)
mlpa:add(prla)

criterion = nn.MarginRankingCriterion(0.1)

x, y, z = torch.randn(5), torch.randn(5), torch.randn(5)

parameters, gradParameters = mlpa:getParameters()
print(parameters:size(), gradParameters:size()) -- show size difference

function feval(params)
    local pred = mlpa:forward({{x, y}, {x, z}})
    local err = criterion:forward(pred, 1)
    local gradCriterion = criterion:backward(pred, 1)
    mlpa:backward({{x, y}, {x, z}}, gradCriterion)
    return err, gradParameters
end
optErr = optim.sgd(feval, parameters, {learningRate=0.01})
```

Which gives me:

```
 12
[torch.LongStorage of size 1]
 48
[torch.LongStorage of size 1]
luajit: .../share/lua/5.1/optim/sgd.lua:81: inconsistent tensor size at /tmp/luarocks_torch-scm-1-5221/torch7/lib/TH/generic/THTensorMath.c:456
stack traceback:
        [C]: in function 'add'
        .../share/lua/5.1/optim/sgd.lua:81: in function 'sgd'
        example.lua:37: in main chunk
        [C]: in function 'dofile'
        .../lib/luarocks/rocks/trepl/scm-1/bin/th:131: in main chunk
        [C]: at 0x00406260
```

The example in the documentation uses `updateParameters()`, which handles the parameter updating differently from optim's addition: see e.g. https://github.com/torch/optim/blob/master/sgd.lua#L81 versus the following in `nn/Module.lua`:

```
for i=1,#params do
   params[i]:add(-learningRate, gradParams[i])
end
```

This is a bit confusing. Of course I could somehow resize the gradParameters in feval, but that doesn't seem to be the right way to do this? I would say that if something works with the simple backward-updateParameters loop it should also work with optim. Am I missing something here?

**Comment by soumith** [comment]
when you are cloning weights/biases, the usual practice is to also clone gradWeight / gradBias

If you do that, you wont have this issue.

**Comment by yjy765** [comment]
@soumith I have a question on sharing. When I shared weight,bias,gradWeight,gradBias, the model performs forward and backward without any problem. The gradient of  each parameter is correctly estimated. However, when I call updateParameters(), each parameter is updated by too much amount. Say, if you share some layer three times, then the parameter of that layer is subtracted by three times of estimated calculate. It is because each layer calls updateParameter though gradient is summed.
How can I solve it?
